docs/proposals/20240102-upgrade-yurtappset-to-v1beta1.md

```diff
@@ -0,0 +1,258 @@
+	// +optional
+	Image *ImageTweak `json:"imageTweak,omitempty"`
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+}
+
+// ImageTweak specifies the corresponding container and the target image
+type ImageTweak struct {
+	// ContainerName represents name of the container in which the Image will be replaced
+	ContainerName string `json:"containerName"`
+	// TargetImage represents the image name which is injected into the container above
+	TargetImage string `json:"targetImage"`
+}
+
+type Operation string
+
+const (
+	ADD     Operation = "add"     // json patch
+	REMOVE  Operation = "remove"  // json patch
+	REPLACE Operation = "replace" // json patch
+)
+
+type AdvancedTweak struct {
+	// Path represents the path in the json patch
+	Path string `json:"path"`
+	// Operation represents the operation
+	// +kubebuilder:validation:Enum=add;remove;replace
+	Operation Operation `json:"operation"`
+	// Indicates the value of json patch
+	// +optional
+	Value apiextensionsv1.JSON `json:"value,omitempty"`
+}
+
+// YurtAppSetStatus defines the observed state of YurtAppSet.
+type YurtAppSetStatus struct {
+	// ObservedGeneration is the most recent generation observed for this YurtAppSet. It corresponds to the
+	// YurtAppSet's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Count of hash collisions for the YurtAppSet. The YurtAppSet controller
+	// uses this field as a collision avoidance mechanism when it needs to
+	// create the name for the newest ControllerRevision.
+	// +optional
+	CollisionCount *int32 `json:"collisionCount,omitempty"`
+
+	// CurrentRevision, if not empty, indicates the current version of the YurtAppSet.
+	CurrentRevision string `json:"currentRevision"`
+
+	// Represents the latest available observations of a YurtAppSet's current state.
+	// +optional
+	Conditions []YurtAppSetCondition `json:"conditions,omitempty"`
+
+	// The number of ready workloads.
+	ReadyWorkloads int32 `json:"readyWorkloads"`
+
+	// The number of updated workloads.
+	// +optional
+	UpdatedWorkloads int32 `json:"updatedWorkloads"`
+
+	// TotalWorkloads is the most recently observed number of workloads.
+	TotalWorkloads int32 `json:"totalWorkloads"`
+}
+
+// YurtAppSetConditionType indicates valid conditions type of a YurtAppSet.
+type YurtAppSetConditionType string
+
+const (
+	// AppDispatched means all the expected workloads are created successfully.
+	AppSetAppDisPatchced YurtAppSetConditionType = "AppDispatched"
+	// AppDeleted means all the unexpected workloads are deleted successfully.
+	AppSetAppDisDeleted YurtAppSetConditionType = "AppDeleted"
+	// AppUpdated means all expected workload are updated successfully.
+	AppSetAppUpdated YurtAppSetConditionType = "AppUpdated"
+	// PoolFound is added to a YurtAppSet when one of owned workloads has failure during its own reconciling.
+	AppSetPoolFound YurtAppSetConditionType = "PoolFound"
+)
```

Comment:
- some other conditions:

AppSetPoolNotFound: no nodepools meets the nodepoolselector of yurtappset
AppSetTemplateInvalid: template in yurtappset is invalid

Reply:
I have different opinions here:
- AppSetPoolNotFound is equivalent to AppSetPoolFound conditionstatus=false.
- it's better to put templateInvalid into the condition reason/message in AppDispatched/AppUpdated condition

Reply:
@luc99hen I agree with your ideas.